src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -226,24 +226,21 @@ private static void CheckEndReachable(object startSegment, object endSegment)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private long GetLength(in SequencePosition start, in SequencePosition end)
         {
-            GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out SequenceType type, out int startIndex, out int endIndex);
-
+            int startIndex = GetIndex(start);
+            int endIndex = GetIndex(end);
             object startObject = start.GetObject();
             object endObject = end.GetObject();
-            if (type == SequenceType.MultiSegment && startObject != endObject)
```

Comment:
I think we need to keep this type check.

We also should probably throw if `type != SequenceType.MultiSegment` and `startObject != endObject`